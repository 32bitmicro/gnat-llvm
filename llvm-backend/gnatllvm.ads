package GNATLLVM is
end GNATLLVM;
