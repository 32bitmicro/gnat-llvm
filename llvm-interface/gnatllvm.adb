------------------------------------------------------------------------------
--                             G N A T - L L V M                            --
--                                                                          --
--                     Copyright (C) 2013-2018, AdaCore                     --
--                                                                          --
-- This is free software;  you can redistribute it  and/or modify it  under --
-- terms of the  GNU General Public License as published  by the Free Soft- --
-- ware  Foundation;  either version 3,  or (at your option) any later ver- --
-- sion.  This software is distributed in the hope  that it will be useful, --
-- but WITHOUT ANY WARRANTY;  without even the implied warranty of MERCHAN- --
-- TABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public --
-- License for  more details.  You should have  received  a copy of the GNU --
-- General  Public  License  distributed  with  this  software;   see  file --
-- COPYING3.  If not, go to http://www.gnu.org/licenses for a complete copy --
-- of the license.                                                          --
------------------------------------------------------------------------------

package body GNATLLVM is

   ------------------------
   -- Get_LLVM_Error_Msg --
   ------------------------

   function Get_LLVM_Error_Msg (Msg : Ptr_Err_Msg_Type) return String is
      Err_Msg_Length : Integer := Msg'Length;
   begin
      for J in Err_Msg_Type'Range loop
         if Msg (J) = ASCII.NUL then
            Err_Msg_Length := J - 1;
            exit;
         end if;
      end loop;

      return Msg (1 .. Err_Msg_Length);
   end Get_LLVM_Error_Msg;

end GNATLLVM;
