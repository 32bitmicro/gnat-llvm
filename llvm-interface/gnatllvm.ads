------------------------------------------------------------------------------
--                             G N A T - L L V M                            --
--                                                                          --
--                     Copyright (C) 2013-2019, AdaCore                     --
--                                                                          --
-- This is free software;  you can redistribute it  and/or modify it  under --
-- terms of the  GNU General Public License as published  by the Free Soft- --
-- ware  Foundation;  either version 3,  or (at your option) any later ver- --
-- sion.  This software is distributed in the hope  that it will be useful, --
-- but WITHOUT ANY WARRANTY;  without even the implied warranty of MERCHAN- --
-- TABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public --
-- License for  more details.  You should have  received  a copy of the GNU --
-- General  Public  License  distributed  with  this  software;   see  file --
-- COPYING3.  If not, go to http://www.gnu.org/licenses for a complete copy --
-- of the license.                                                          --
------------------------------------------------------------------------------

with System;

with stdint_h; use stdint_h;

with Interfaces.C;
with Interfaces.C.Extensions;

with Atree;    use Atree;
with Einfo;    use Einfo;
with Namet;    use Namet;
with Types;    use Types;
with Uintp;    use Uintp;

with LLVM.Target;         use LLVM.Target;
with LLVM.Target_Machine; use LLVM.Target_Machine;
with LLVM.Types;          use LLVM.Types;

package GNATLLVM is

   --  Note: in order to find the right LLVM instruction to generate,
   --  you can compare with what Clang generates on corresponding C or C++
   --  code. This can be done online via http://ellcc.org/demo/index.cgi

   --  See also DragonEgg sources for comparison on how GCC nodes are converted
   --  to LLVM nodes: http://llvm.org/svn/llvm-project/dragonegg/trunk

   --  This package contains very low-level types, objects, and functions,
   --  mostly corresponding to LLVM objects and exists to avoid circular
   --  dependencies between other specs.  The intent is that every child
   --  package of LLVM with's this child, but that this with's no other
   --  children.

   --  Define unsigned and unsigned_long_long here instead of use'ing
   --  Interfaces.C and Interfaces.C.Extensions because the latter brings
   --  in a lot of junk that gets in the way and the former conflicts with
   --  Int from the front end.

   subtype unsigned is Interfaces.C.unsigned;
   function "+" (L, R : unsigned)  return unsigned renames Interfaces.C."+";
   function "-" (L, R : unsigned)  return unsigned renames Interfaces.C."-";
   function "*" (L, R : unsigned)  return unsigned renames Interfaces.C."*";
   function "/" (L, R : unsigned)  return unsigned renames Interfaces.C."/";
   function "=" (L, R : unsigned)  return Boolean  renames Interfaces.C."=";
   function ">" (L, R : unsigned)  return Boolean  renames Interfaces.C.">";
   function "<" (L, R : unsigned)  return Boolean  renames Interfaces.C."<";
   function "<=" (L, R : unsigned) return Boolean  renames Interfaces.C."<=";
   function ">=" (L, R : unsigned) return Boolean  renames Interfaces.C.">=";

   subtype unsigned_long_long is Interfaces.C.Extensions.unsigned_long_long;
   subtype ULL is unsigned_long_long;
   --  Define shorter alias name for easier reading

   function "+"  (L, R : ULL)  return ULL renames Interfaces.C.Extensions."+";
   function "-"  (L, R : ULL)  return ULL renames Interfaces.C.Extensions."-";
   function "*"  (L, R : ULL)  return ULL renames Interfaces.C.Extensions."*";
   function "/"  (L, R : ULL)  return ULL renames Interfaces.C.Extensions."/";

   function "**" (L : ULL; R : Integer) return ULL
     renames Interfaces.C.Extensions."**";
   function "mod"  (L, R : ULL)  return ULL
     renames Interfaces.C.Extensions."mod";
   function "and"  (L, R : ULL)  return ULL
     renames Interfaces.C.Extensions."and";

   function "="  (L, R : ULL)  return Boolean
     renames Interfaces.C.Extensions."=";
   function ">"  (L, R : ULL)  return Boolean
     renames Interfaces.C.Extensions.">";
   function "<"  (L, R : ULL)  return Boolean
     renames Interfaces.C.Extensions."<";
   function "<=" (L, R : ULL) return  Boolean
     renames Interfaces.C.Extensions."<=";
   function ">=" (L, R : ULL) return  Boolean
     renames Interfaces.C.Extensions.">=";

   subtype long_long_integer is Interfaces.C.Extensions.long_long;
   subtype LLI is long_long_integer;
   --  Define shorter alias name for easier reading

   function UI_From_LLI is new UI_From_Integral (LLI);
   --  And conversion routine from it to UI

   type MD_Builder_T is new System.Address;
   --  Metadata builder type: opaque for us

   type Value_Array        is array (Nat range <>) of Value_T;
   type Basic_Block_Array  is array (Nat range <>) of Basic_Block_T;
   type Index_Array        is array (Nat range <>) of unsigned;
   type Access_Value_Array is access all Value_Array;

   No_Value_T    : constant Value_T       := Value_T (System.Null_Address);
   No_Type_T     : constant Type_T        := Type_T (System.Null_Address);
   No_BB_T       : constant Basic_Block_T :=
     Basic_Block_T (System.Null_Address);
   No_Metadata_T : constant Metadata_T    := Metadata_T (System.Null_Address);
   No_Builder_T  : constant Builder_T     := Builder_T (System.Null_Address);
   No_Module_T   : constant Module_T      := Module_T (System.Null_Address);
   --  Constant for null objects of various types

   function No (V : Value_T)            return Boolean is (V = No_Value_T);
   function No (T : Type_T)             return Boolean is (T = No_Type_T);
   function No (B : Basic_Block_T)      return Boolean is (B = No_BB_T);
   function No (M : Metadata_T)         return Boolean is (M = No_Metadata_T);
   function No (M : Builder_T)          return Boolean is (M = No_Builder_T);
   function No (N : Name_Id)            return Boolean is (N = No_Name);

   function Present (V : Value_T)       return Boolean is (V /= No_Value_T);
   function Present (T : Type_T)        return Boolean is (T /= No_Type_T);
   function Present (B : Basic_Block_T) return Boolean is (B /= No_BB_T);
   function Present (M : Metadata_T)    return Boolean is (M /= No_Metadata_T);
   function Present (M : Builder_T)     return Boolean is (M /= No_Builder_T);
   function Present (N : Name_Id)       return Boolean is (N /= No_Name);
   --  Test for presence and absence of fields of LLVM types

   function No      (U : Uint) return Boolean is (U = No_Uint);
   function Present (U : Uint) return Boolean is (U /= No_Uint);
   --  Add presence and absense functions for Uint for consistency

   function Is_Type_Or_Void (E : Entity_Id) return Boolean is
     (Ekind (E) = E_Void or else Is_Type (E));
   --  We can have Etype's that are E_Void for E_Procedure

   --  We have to define GL_Type here so it can be used in GNATLLVM.GLValue,
   --  which GNATLLVM.GLType must reference.

   GL_Type_Low_Bound  : constant := 700_000_000;
   GL_Type_High_Bound : constant := 799_999_999;
   type GL_Type is range GL_Type_Low_Bound .. GL_Type_High_Bound;
   No_GL_Type         : constant GL_Type := GL_Type_Low_Bound;

   function No      (GT : GL_Type) return Boolean is (GT =  No_GL_Type);
   function Present (GT : GL_Type) return Boolean is (GT /= No_GL_Type);

   type Word_Array is array (Nat range <>) of aliased uint64_t;
   --  Array of words for LLVM construction functions

   Context            : Context_T;
   --  The current LLVM Context

   IR_Builder         : Builder_T;
   --  The current LLVM Instruction builder

   DI_Builder          : DI_Builder_T;
   --  The current LLVM Debug Info builder

   Module             : Module_T;
   --  The LLVM Module being compiled

   LLVM_Target        : Target_T;
   --  The LLVM target for our module

   Target_Machine     : Target_Machine_T;
   --  The LLVM target machine for our module

   MD_Builder         : MD_Builder_T;
   --  The current LLVM Metadata builder

   TBAA_Root          : Metadata_T;
   --  Root of tree for Type-Based alias Analysis (TBAA) metadata

   Module_Data_Layout : Target_Data_T;
   --  LLVM current module data layout.

   Convert_Module     : Module_T;
   --  The module use by Convert_Nonsymbolic_Constant

   Size_Type          : Entity_Id := Empty;
   Size_GL_Type       : GL_Type   := No_GL_Type;
   LLVM_Size_Type     : Type_T    := No_Type_T;
   --  Types to use for sizes

   Void_Ptr_Type      : Type_T;
   --  Pointer to arbitrary memory (we use i8 *); equivalent of
   --  Standard_A_Char.

   A_Char_GL_Type    : GL_Type;
   Boolean_GL_Type   : GL_Type;
   SSI_GL_Type       : GL_Type;
   SI_GL_Type        : GL_Type;
   Integer_GL_Type   : GL_Type;
   LI_GL_Type        : GL_Type;
   LLI_GL_Type       : GL_Type;
   Void_GL_Type      : GL_Type;
   Any_Array_GL_Type : GL_Type;
   --  GL_Types for builtin types

   Int_32_GL_Type    : GL_Type;
   Int_32_Type       : Entity_Id;
   --  Type for 32-bit integers (for GEP indexes)

   BPU                : Int;
   Bit_T              : Type_T;
   Byte_T             : Type_T;

   --  GNAT LLVM switches

   Decls_Only      : Boolean        := False;
   --  True if just compiling to process and back-annotate decls

   Output_Assembly : Boolean := False;
   --  True if -S was specified

   Emit_LLVM       : Boolean := False;
   --  True if -emit-llvm was specified

   Emit_Debug_Info : Boolean := False;
   --  Whether or not to emit debugging information (-g)

   Do_Stack_Check  : Boolean := False;
   --  If set, check for too-large allocation

   Short_Enums     : Boolean := False;
   --  True if we should use the RM_Size, not Esize, for enums

   subtype Err_Msg_Type is String (1 .. 10000);
   type Ptr_Err_Msg_Type is access all Err_Msg_Type;
   --  Used for LLVM error handling

end GNATLLVM;
