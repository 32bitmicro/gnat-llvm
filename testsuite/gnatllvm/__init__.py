from collections import namedtuple
import ctypes
from ctypes import cdll
import os
import os.path
import subprocess
import sys

from gnatpython.fileutils import mkdir, rm, which


TESTSUITE_ROOT = os.environ['TESTSUITE_ROOT']

GNATLLVM_GPR = os.path.join(TESTSUITE_ROOT, '..', 'gnat_llvm.gpr')
SRCCOV_DIR = os.path.join(TESTSUITE_ROOT, 'srccov')
UNITS_LIST = os.path.join(SRCCOV_DIR, 'units.list')
TRACES_DIR = os.path.join(SRCCOV_DIR, 'traces')

USE_NATIVE_GNAT = bool(os.environ.get('USE_NATIVE_GNAT', False))
SOURCE_COVERAGE_LEVEL = os.environ.get('SOURCE_COVERAGE_LEVEL', None)

LLVM_GNATCOMPILE = which('llvm-gcc')
TESTCASE_NAME = os.environ['TESTCASE_NAME']

Func = namedtuple('Func', 'name argtypes restype')


trace_count = 0

def get_trace_filename():
    global trace_count
    trace_count += 1
    test_basename = TESTCASE_NAME.lstrip('./').replace('/', '-')
    return '{}-{}.trace'.format(test_basename, trace_count)

def get_library_name(filename):
    """Return the library file name for the given Ada source file name."""
    return os.path.splitext(filename)[0]

def change_ext(filename, new_ext):
    """Replace "filename" extension with "new_ext"."""
    return '{}{}{}'.format(
        os.path.splitext(filename)[0],
        os.path.extsep,
        new_ext
    )


def get_shared_func(shared, name, argtypes=[], restype=None):
    """Get a function from a shared object.

    Return the "name" function from the "shared" shared object. Also specify
    argument types and result type.
    """
    func = getattr(shared, name)
    func.argtypes = argtypes
    func.restype = restype
    return func


def gnat_to_bc(adb, gargs=None):
    """Compile the "adb" unit and return the result bitcode filename."""

    if USE_NATIVE_GNAT:
        raise RuntimeError(
            'Cannot produce LLVM bitcode with the native GNAT compiler')

    prefix = (
        [] if not SOURCE_COVERAGE_LEVEL else
        [
            'gnatcov', 'run',
            '-o', os.path.join(TRACES_DIR, get_trace_filename()),
            '-c', SOURCE_COVERAGE_LEVEL,
            '-P{}'.format(GNATLLVM_GPR),
            '--units=@{}'.format(UNITS_LIST),
            '-eargs',
        ]
    )
    gargs = list(gargs) if gargs else []
    subprocess.check_call(prefix + [LLVM_GNATCOMPILE, '-c', adb] + gargs)
    return change_ext(adb, 'bc')

def gnat_to_obj(adb, gargs=None):
    """Compile the "adb" unit and return the result object file filename."""

    if USE_NATIVE_GNAT:
        gargs = list(gargs) if gargs else []
        subprocess.check_call(['gcc', '-c', '-fPIC', '-g'] + gargs + [adb])
    else:
        subprocess.check_call(
            ['llc', '-relocation-model=pic', gnat_to_bc(adb, gargs)])
        subprocess.check_call(['gcc', '-c', change_ext(adb, 's')])
    return change_ext(adb, 'o')

def gnat_to_shared(adb_list, name, gargs=None):
    """Compile sources to a shared object.

    Compile the set of unit "adb_list" and link them to a dynamic library
    "name". Return the shared object filename.
    """
    obj_list = [gnat_to_obj(adb, gargs) for adb in adb_list]
    result = change_ext(name, 'so')

    if USE_NATIVE_GNAT:
        library_list = [get_library_name(adb) for adb in adb_list]
        binder_output = 'b~{}.adb'.format(library_list[-1])
        subprocess.check_call(
            ['gnatbind', '-L{}'.format(name), '-shared'] + library_list)
        obj_list.append(gnat_to_obj(binder_output, gargs))
        extra_options = ['-lgnat']
    else:
        extra_options = []

    subprocess.check_call(
        ['gcc', '-shared', '-o', result] + extra_options + obj_list)
    return result

def build_and_load(adb_list, name, *objects, **kwargs):
    """Build a shared object from sources and load objects from it.

    Compile the set of unit "adb_list" and link them to a dynamic library
    "name". Then, load it using ctypes and return the sequence of loaded
    "objects".
    """
    shared = gnat_to_shared(adb_list, name, **kwargs)
    lib = cdll.LoadLibrary(shared)
    result = []

    for obj in objects:
        if isinstance(obj, Func):
            loaded_obj = get_shared_func(
                lib, obj.name, obj.argtypes, obj.restype)
        else:
            raise TypeError('Invalid object type: {}'.format(type(obj)))
        result.append(loaded_obj)

    return result


class Structure(ctypes.Structure):

    def __str__(self):
        return '{} ({})'.format(type(self).__name__, ', '.join(
            '{}={}'.format(field_name, getattr(self, field_name))
            for field_name, field_type in self._fields_
        ))
