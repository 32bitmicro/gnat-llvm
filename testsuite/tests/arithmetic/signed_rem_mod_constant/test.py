from ctypes import *
from gnatllvm import build_and_load, Func

(Foo, ) = build_and_load(
    ['foo.adb'], 'foo',
    Func('_ada_foo', argtypes=[c_bool, c_bool, c_int], restype=c_int),
)

for modrem, sign, a in (
    (True, False, -6),
    (True, False, -5),
    (True, False, -4),
    (True, False, -1),
    (True, False, 0),
    (True, False, 1),
    (True, False, 4),
    (True, False, 5),
    (True, False, 6),
    (True, True, -6),
    (True, True, -5),
    (True, True, -4),
    (True, True, -1),
    (True, True, 0),
    (True, True, 1),
    (True, True, 4),
    (True, True, 5),
    (True, True, 6),
    (False, False, -6),
    (False, False, -5),
    (False, False, -4),
    (False, False, -1),
    (False, False, 0),
    (False, False, 1),
    (False, False, 4),
    (False, False, 5),
    (False, False, 6),
    (False, True, -6),
    (False, True, -5),
    (False, True, -4),
    (False, True, -1),
    (False, True, 0),
    (False, True, 1),
    (False, True, 4),
    (False, True, 5),
    (False, True, 6),
):
    print 'Foo ({}, {}, {}) = {}'.format(modrem, sign, a, Foo(modrem, sign, a))
