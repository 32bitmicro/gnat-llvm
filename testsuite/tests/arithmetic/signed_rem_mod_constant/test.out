Foo (True, False, -6) = 4
Foo (True, False, -5) = 0
Foo (True, False, -4) = 1
Foo (True, False, -1) = 4
Foo (True, False, 0) = 0
Foo (True, False, 1) = 1
Foo (True, False, 4) = 4
Foo (True, False, 5) = 0
Foo (True, False, 6) = 1
Foo (True, True, -6) = -1
Foo (True, True, -5) = 0
Foo (True, True, -4) = -4
Foo (True, True, -1) = -1
Foo (True, True, 0) = 0
Foo (True, True, 1) = -4
Foo (True, True, 4) = -1
Foo (True, True, 5) = 0
Foo (True, True, 6) = -4
Foo (False, False, -6) = -1
Foo (False, False, -5) = 0
Foo (False, False, -4) = -4
Foo (False, False, -1) = -1
Foo (False, False, 0) = 0
Foo (False, False, 1) = 1
Foo (False, False, 4) = 4
Foo (False, False, 5) = 0
Foo (False, False, 6) = 1
Foo (False, True, -6) = -1
Foo (False, True, -5) = 0
Foo (False, True, -4) = -4
Foo (False, True, -1) = -1
Foo (False, True, 0) = 0
Foo (False, True, 1) = 1
Foo (False, True, 4) = 4
Foo (False, True, 5) = 0
Foo (False, True, 6) = 1
